static/js/components/forms/B2BPurchaseForm.js

```diff
@@ -35,89 +42,152 @@ export const validate = (values: Object) => {
   return errors
 }
 
-const B2BPurchaseForm = ({ onSubmit, products, requestPending }: Props) => (
-  <Formik
-    onSubmit={onSubmit}
-    initialValues={{
-      num_seats: "",
-      email:     "",
-      product:   ""
-    }}
-    validate={validate}
-    render={({ values }) => {
-      let itemPrice, totalPrice, numSeats
-      const productId = parseInt(values.product)
-      const product = products.find(product => product.id === productId)
-      const productVersion = product ? product.latest_version : null
-      if (productVersion) {
-        numSeats = parseInt(values.num_seats)
-        if (!isNaN(numSeats) && productVersion.price !== null) {
-          itemPrice = new Decimal(productVersion.price)
-          totalPrice = itemPrice * new Decimal(numSeats)
+class B2BPurchaseForm extends React.Component<Props> {
+  applyCoupon = async (
+    event: Event,
+    values: Object,
+    setFieldError: Function
+  ) => {
+    const { clearCouponStatus, fetchCouponStatus } = this.props
+
+    event.preventDefault()
+
+    if (!values.coupon) {
+      clearCouponStatus()
+      return
+    }
+
+    if (!values.product) {
+      setFieldError("coupon", "No product selected")
+      return
+    }
+
+    const response = await fetchCouponStatus({
+      product_id: values.product,
+      code:       values.coupon
+    })
+    if (response.status !== 200) {
+      setFieldError("coupon", "Invalid coupon code")
+    }
+  }
+
+  renderForm = ({ values, setFieldError }: Object) => {
+    const { products, requestPending, couponStatus } = this.props
+
+    let itemPrice, totalPrice, discount
+    const productId = parseInt(values.product)
+    const product = products.find(product => product.id === productId)
+    const productVersion = product ? product.latest_version : null
+    const numSeats = parseInt(values.num_seats)
+    if (productVersion && productVersion.price !== null) {
+      itemPrice = new Decimal(productVersion.price)
+      if (!isNaN(numSeats)) {
+        totalPrice = itemPrice.times(numSeats)
+
+        if (couponStatus) {
+          discount = new Decimal(couponStatus.discount_percent)
+            .times(itemPrice)
+            .times(numSeats)
+          totalPrice = totalPrice.minus(discount)
         }
       }
+    }
```

Comment:
Sorry I didn't bring this up in the last review, but there's something a bit awkward about running through all this logic to calculate these values in a form render function, then simply passing these values down to a different component that actually needs them. Do you see a clean way to reimplement this so that `B2BPurchasePage` renders the main grid elements, `B2BPurchaseForm` just renders the contents of the `col-lg-5` column, and `B2BPurchasePage` renders the `B2BPurchaseSummary` directly?

```javascript
// B2BPurchasePage
render() {
      <div className="container">
        <div className="row">
          <div className="col-lg-12">
            <div className="title">Bulk Seats</div>
          </div>
        </div>
        <div className="row">
          <div className="col-lg-5">
            <B2BPurchaseForm ... />
          </div>
          <div className="col-lg-3" />
          <div className="col-lg-4">
            <B2BPurchaseSummary ... />
            <button>...</button>
          </div>
          //...
}
```

This calculation for these variables in the lines above could also be done in `B2BPurchasePage`. If all of that doesn't seem feasible or would require a ton of work, feel free to skip.

Reply:
I don't think that plan could work because these lines depend on the form values so it would need to be done in `B2BPurchaseForm` or another component inside it.